Libunwind1.5rc2

sdmaclea opened the pull request:
If we want cross DAC to push changes to upstream the first step would be to update to latest upstream.

Keeps the files .NET added.
```bash
ls -laR
.:
total 16
drwxrwxr-x  3 stmaclea stmaclea 4096 May  7 17:26 .
drwxrwxr-x 28 stmaclea stmaclea 4096 May  7 15:35 ..
-rw-rw-r--  1 stmaclea stmaclea  645 May  7 17:26 CMakeLists.txt
lrwxrwxrwx  1 stmaclea stmaclea    6 May  7 17:04 README.md -> README
drwxrwxr-x  3 stmaclea stmaclea 4096 May  7 17:26 src

./src:
total 36
drwxrwxr-x 3 stmaclea stmaclea  4096 May  7 17:26 .
drwxrwxr-x 3 stmaclea stmaclea  4096 May  7 17:26 ..
-rw-rw-r-- 1 stmaclea stmaclea 13740 May  7 16:59 CMakeLists.txt
-rw-rw-r-- 1 stmaclea stmaclea   340 May  7 16:59 config.h.in
-rw-rw-r-- 1 stmaclea stmaclea   804 May  7 16:59 configure.cmake
drwxrwxr-x 2 stmaclea stmaclea  4096 May  7 16:59 oop

./src/oop:
total 16
drwxrwxr-x 2 stmaclea stmaclea 4096 May  7 16:59 .
drwxrwxr-x 3 stmaclea stmaclea 4096 May  7 17:26 ..
-rw-rw-r-- 1 stmaclea stmaclea 3036 May  7 16:59 _OOP_find_proc_info.c
-rw-rw-r-- 1 stmaclea stmaclea  534 May  7 16:59 _OOP_internal.h
```

It discards local changes to the source.
dotnet/coreclr#26869 (Should be upstreamed)
dotnet/coreclr#26108 (Unimportant)

sdmaclea committed:
Add libunwind 1.5rc2 source

Rename unused autoconfig dir aux -> aux_ consistenet with
original checkin

sdmaclea committed:
Delete files added by 1.3-rc1

sdmaclea committed:
Update libunwind version in CMake

janvorli reviewed (approved):
LGTM, thank you for updating it!

janvorli commented:
cc @am11

janvorli commented:
@sdmaclea I don't see any changes in the original libunwind in  dotnet/coreclr#27795. All the cmake projects are ours.

janvorli commented:
As for the change dotnet/coreclr#26108, it should have not been made anyways.

sdmaclea commented:
The actual change was dotnet/coreclr#26869.  It was insignificant for us.

janvorli commented:
> The actual change was [dotnet/coreclr#26869](https://github.com/dotnet/coreclr/pull/26869). It was insignificant for us.

Seems like it would make sense to upstream it too for the benefit of the community, as that assert was incorrectly placed.

am11 commented:
I was waiting for 5.0 final release before pushing my branch to PR https://github.com/am11/runtime/commit/6bf7a1bde7c54524fa5177e263c435ca9394c257.

@sdmaclea, if we are going ahead with it, could you please add a `src/coreclr/src/pal/src/libunwind/libunwind-version.txt` file similar to https://github.com/dotnet/runtime/blob/master/src/libraries/Native/AnyOS/brotli-version.txt?

Also, if possible, please cherry-pick changes from my branch in the following two files: `src/coreclr/src/pal/src/CMakeLists.txt` and `src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt`; otherwise, I will do a followup PR.

janvorli commented:
Btw, the build is failing due to 
```
2020-05-07T21:30:25.9173704Z /__w/1/s/src/coreclr/src/pal/src/libunwind/src/dwarf/Gfind_proc_info-lsb.c:61:9: error: 'dwarf_search_unwind_table_int' macro redefined [-Werror,-Wmacro-redefined]
2020-05-07T21:30:25.9180928Z #define dwarf_search_unwind_table_int dwarf_search_unwind_table
2020-05-07T21:30:25.9235183Z         ^
2020-05-07T21:30:25.9236066Z <command line>:17:9: note: previous definition is here
2020-05-07T21:30:25.9236990Z #define dwarf_search_unwind_table_int UNW_OBJ(dwarf_search_u
```

The fix should be to remove the definition of the dwarf_search_unwind_table_int from here:
https://github.com/dotnet/runtime/blob/0ec974709e311ef63e7e23ed7c74f1f33762aecc/src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt#L18

sdmaclea committed:
Add libunwind-version.txt

am11 committed:
Add changes for SunOS from @am11

sdmaclea committed:
Revert change to oop

sdmaclea committed:
Remove obsolete add_definition

am11 commented on src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt:
```diff
@@ -14,8 +14,6 @@ add_definitions(-DPACKAGE_STRING="")
 
 add_definitions(-D_GNU_SOURCE)
-# Ensure that the remote and local unwind code can reside in the same binary without name clashing
-add_definitions("-Ddwarf_search_unwind_table_int=UNW_OBJ(dwarf_search_unwind_table_int)")
```
Based on https://github.com/libunwind/libunwind/issues/176, this will fix musl arm64 build:

```cmake
if (CLR_CMAKE_TARGET_ALPINE_LINUX AND CLR_CMAKE_TARGET_ARCH_ARM64)
  add_definitions("-D__sigset_t=sigset_t")
endif()
```

am11 commented:
>  dotnet/coreclr#26108 (Unimportant)

That patch was actually a bug. The interpreter specified in autogen.sh is POSIX shell (uses `#!/bin/sh` shebang in first line) and patch uses `${BASH_SOURCE[0]}`: a) POSIX shell does not define arrays, b) BASH_SOURCE is bash-specific thing. We do not use autogen.sh during the build, that's why breakage was not visible.

sdmaclea commented:
Ack.  It was unimportant for us, since we don't use the file.

sdmaclea committed:
Fix musl build

am11 commented on src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt:
```diff
@@ -51,6 +51,10 @@ elseif(CLR_CMAKE_HOST_ARCH_ARM64)
         add_compile_options(-Wno-absolute-value)
     endif()
+    // Issue https://github.com/libunwind/libunwind/issues/176
+    if (CLR_CMAKE_TARGET_ALPINE_LINUX)
```
it only applies to arm64:

```suggestion
    if (CLR_CMAKE_TARGET_ALPINE_LINUX AND CLR_CMAKE_TARGET_ARCH_ARM64)
```

sdmaclea replied:
It already in a ARM64 block.

sdmaclea committed:
Fix comment

sdmaclea committed:
Be consistent and use HOST

hoyosjs commented:
We had made some changes to libunwind (arm out of process unwind for dump collection if I remember correctly @mikem8361) and there's issue https://github.com/dotnet/runtime/issues/2014 tracking the need to use stock libunwind for RHEL sourcebuild. Does this conflict with any of these two?

am11 commented:
> arm out of process unwind for dump collection

`oop/_OOP_find_proc_info.c` and friends are kept intact by this PR.

> libunwind for RHEL sourcebuild

https://github.com/dotnet/runtime/issues/2014 refers to the missing symbol issue, when we use distro provided libunwind.so and skip the in-tree compilation. So this PR has no effect on that situation.

sdmaclea commented:
Looks like this is update is breaking `arm64`.  I'll take a look tomorrow.

sdmaclea committed:
Fix error in unw_sigcontext

Introduced by libunwind/libunwind#71

__reseverved needs to be big enough to store a unw_fpsimd_context_t
Which includes 32 128-bit registers, stored as 64 64-bit half registers.
Fix off by 2 issue

sdmaclea committed:
Arm64 support !UNWIND_CONTEXT_IS_UCONTEXT_T

janvorli commented on src/coreclr/src/pal/src/libunwind/include/libunwind-aarch64.h:
```diff
@@ -184,7 +184,7 @@ struct unw_sigcontext
 	uint64_t pc;
 	uint64_t pstate;
-	uint8_t __reserved[(34 * 8)] __attribute__((__aligned__(16)));
+	uint8_t __reserved[(66 * 8)] __attribute__((__aligned__(16)));
```
Where was this causing a problem? The comment above says this field was intentionally trimmed down.

sdmaclea replied:
I pushed the patch upstream.  It merged today as libunwind/libunwind#179
It was trimmed down from 4096.
It needed to be 66 to match unw_fpsimd_t below.

safern commented:
@sdmaclea this is causing libraries test to crash on Linux arm64. Unfortunately we don't have a dump, it seems like linux jobs are not uploading dumps (maybe only jobs inside docker are failing to do so?), @missymessa do you know how can we get one?

In the meantime I put up: #36909 to revert this change.
